### PR TITLE
Fix issue #76 by using deep clone

### PR DESCRIPTION
**Pull request by baranga**
This is the same as PR #100 but based on the latest release. This only fixes the issue for core and not for duplex.

As @tomalec already mentioned in #100 this comes with a speed penalty. A quick check showed a run time increase of more then 100% on my local machine.

### CODE REVIEW

**Comment by tomalec** [comment]
Thanks a ton for such a quick rebase! :)

Some tests are still failing https://travis-ci.org/Starcounter-Jack/JSON-Patch/builds/184538198 could you check it.

I have just added CLI benchmarks (`npm run bench`). It would be great if you could rebase/merge your PR to latest master and check how it affects stats.
Please, give me  few mins to release it as a new version, I'll ping you here.

If there is a problem with regular apply it would be good to have a test for it as well.

I'll try to dig today to improve performance of a fix somehow.

**Comment by tomalec** [comment]
ping, 1.1.3 released with `npm run bench`

**Comment by baranga** [comment]
The duplex stuff is actually failing because introducing deep clone there had side effects. I'm not sure how fast I can fix the issue there without breaking the old tests.

**Comment by felixfbecker** [comment]
I am heavily against this. Cloning an object has a considerable performance impact. If you want to ensure the values are not modified, why are you not cloning the patch before passing it to `apply`?

**Comment by felixfbecker** [comment]
Stringify + Parse is also way slower than copying an object by traversing
https://jsperf.com/deep-copy-vs-json-stringify-json-parse/15

**Comment by baranga** [comment]
@felixfbecker I use `validate` to check a request (which calls `apply` internally). This should in no way modify the patch set.

**Comment by felixfbecker** [comment]
I agree, but your PR changes `apply`, not just `validate`, to always clone.

**Comment by baranga** [comment]
Consider this case:
```
var ps = [
  {op: 'add', path: '/foo', value: []},
  {op: 'add', path: '/foo/-', value: 1}
];
var t1 = {};
var t2 = {};

jsonpatch.apply(t1, ps);
jsonpatch.apply(t2, ps);

t1.foo; // [1, 1]
t1.foo.length === 2; // true
t2.foo; // [1, 1]
t2.foo.length === 2; // true
t1.foo === t2.foo; // true
```
Looks pretty broken to me.

**Comment by baranga** [comment]
Quickly integrated traversing for deep clone but performance impact is not that big:
```
Vanilla
==================
Benchmark results:
------------------
add operation
 Ops/sec: 2563197 ±1.32% Ran 137010 times in 0.053 seconds.
remove operation
 Ops/sec: 1143785 ±1.34% Ran 61369 times in 0.054 seconds.
replace operation
 Ops/sec: 3112448 ±1.91% Ran 182727 times in 0.059 seconds.
move operation
 Ops/sec: 377092 ±1.48% Ran 19978 times in 0.053 seconds.
copy operation
 Ops/sec: 748330 ±3.82% Ran 41684 times in 0.056 seconds.
test operation
 Ops/sec: 906874 ±3.09% Ran 51185 times in 0.056 seconds.
===================

Traversing
==================
Benchmark results:
------------------
add operation
 Ops/sec: 677336 ±0.96% Ran 35582 times in 0.053 seconds.
remove operation
 Ops/sec: 1187174 ±1.59% Ran 63270 times in 0.053 seconds.
replace operation
 Ops/sec: 732505 ±1.85% Ran 40863 times in 0.056 seconds.
move operation
 Ops/sec: 202986 ±2.17% Ran 11093 times in 0.055 seconds.
copy operation
 Ops/sec: 291365 ±1.53% Ran 15517 times in 0.053 seconds.
test operation
 Ops/sec: 989830 ±1.08% Ran 52228 times in 0.053 seconds.
===================

JSON
==================
Benchmark results:
------------------
add operation
 Ops/sec: 562641 ±7.57% Ran 36249 times in 0.064 seconds.
remove operation
 Ops/sec: 1086974 ±5.62% Ran 63459 times in 0.058 seconds.
replace operation
 Ops/sec: 686625 ±1.55% Ran 37534 times in 0.055 seconds.
move operation
 Ops/sec: 202023 ±1.64% Ran 11026 times in 0.055 seconds.
copy operation
 Ops/sec: 280681 ±2.38% Ran 15118 times in 0.054 seconds.
test operation
 Ops/sec: 981328 ±1.48% Ran 51489 times in 0.052 seconds.
===================
```

All done with node 7.5.0 and `npm run bench-core`.

**Comment by baranga** [comment]
The high performance of `add` and `replace` is archived by copy by reference instead of copy by value which I consider a fundamental flaw (see above). So I guess comparing `vanilla` against something "correct" isn't fair :)

**Comment by felixfbecker** [comment]
why is this broken? The first operation in your `ps` array has a value that is an array. An array is an object, so it is a reference. The second operation mutates that array. 
So when you apply the patch once to `t1`, it will set `t1.foo` to that array reference, the append an item to it.
When you apply the patch again to `t2`, it will also set `t2.foo` to that array reference (which is now length 1) and append another item to it. At the time you apply the patch to `t2`, it's not the sam e patch anymore. The value set in the first operation is now an array of length 1.
I wouldn't consider this broken at all. If you don't want that behaviour, just clone the patch before applying, it's a one-liner. But other many use cases only apply the patch once and then throw it away, and for those cloning is unneeded overhead.
I don't know how big the values are in the benchmark, but the bigger they are, the greater the overhead.

**Comment by baranga** [comment]
Broken interfaces are failures of the user? I expect a method/function to not modify any arguments if it's not documented and it also returns a result value.

Modification of `obj` parameter is obvious cause that's what `apply` is about. But modification of `patches` isn't documented plus `apply` returns a result array.

**Comment by felixfbecker** [comment]
I agree this could be documented better. But the very essence of `apply` is that it mutates its arguments. You pass it an object to mutate with an array of operations that describe the mutations. If one is "set this property to this array" and the next one is "now append a value to that array", of course that array is mutated, because at the time of the second operation application, it is now part of the document to mutate.

**Comment by baranga** [comment]
Whats about this:
```
var ps = [
  {op:'add', path:'/foo', value: []},
  {op:'add', path:'/foo/-', value:1},
  {op:'copy', from:'/foo', path:'/bar'},
  {op:'add', path:'/bar/-', value:2}
];
var o = {};

jsonpatch.apply(o, ps);
```
I would expect `o` is this: `{foo: [1], bar: [1, 2]}` but it will be this: `{foo: [1, 2], bar: [1, 2]}`.

**Comment by alshakero** [comment]
Yes, I agree with you [this time](https://github.com/Starcounter-Jack/JSON-Patch/issues/177).

**Comment by felixfbecker** [comment]
Hmm, `copy` is a tricky semantic, the spec just says this:

>  This operation is functionally identical to an "add" operation at the
>  target location using the value specified in the "from" member.

**Comment by baranga** [comment]
I find this not that tricky. It's called `copy` and not `link`. Normally one gets two separate things copying after coping.

**Comment by baranga** [comment]
@tomalec See [here](https://github.com/Starcounter-Jack/JSON-Patch/pull/134#issuecomment-303370180) for performance comparison.

**Comment by felixfbecker** [comment]
You can imply that for `move`, but definitely not for `add`

**Comment by baranga** [comment]
@felixfbecker  We talk about JSON-Patch. JSON doesn't know references.

So maybe that's the confusing part: passing a JavaScript value for `patches` instead of a JSON value. I assume `patches` parameter is handled like a plain string (which doesn't get modified), so I can reuse my parameter as often as I want. So for me `value: []` means: use a empty array and not use exactly this array.

**Comment by baranga** [comment]
Changed `_deepClone` to used algorithm #176 and @felixfbecker mentioned for deep clone. Added 2 `apply` test cases for `copy` side effects and unmodified patch sets.

**Comment by baranga** [comment]
Applied the fixes also to `duplex`. All tests green now.

Should I replace all `JSON.parse(JSON.stringify(..))` call with `_deepClone`?

**Comment by felixfbecker** [comment]
Can't we have an option for this? For my use case I have to apply thousands of `add` operations with highly nested values, fast, and don't care about the patch objects at all afterwards. This will have no doubt have a negative performance impact (I don't really trust the benchmark to highlight the worst case scenario here).

**Comment by baranga** [comment]
@felixfbecker at least `copy` should be fixed with deep clone for your case. I can imagine the benchmark doesn't covers your use case. Could you provide any insights so we could hack a custom benchmark?

In general I prefer a slower correct thing over a broken fast one.

**Comment by tomalec** [comment]
I would say we definitely need to fix `copy` and clone there.

Speaking of other `op`s. I have the doubt similar to @felixfbecker whether it's a bug or a feature.

>We talk about JSON-Patch. JSON doesn't know references.

Maybe we could then answer: Then provide JSON as input not a JS object with references:
```js
var ps = `[
  {"op": "add", "path": "/foo", value: []},
  {"op": "add", "path": "/foo/-", value: 1}
]`;
var t1 = {};
var t2 = {};

jsonpatch.apply(t1, ps); // jsonpatch.apply(t1, JSON.parse(ps));
jsonpatch.apply(t2, ps); // jsonpatch.apply(t2, JSON.parse(ps))

t1.foo; // [1, 1]
t2.foo; // [1, 1]
t1.foo === t2.foo; // false
```

That's why I would go for optional flag to fine tune the performance vs. immutability of input. Such flag was proposed by @KamranAsif at https://github.com/Starcounter-Jack/JSON-Patch/issues/157#issuecomment-300875985, so maybe we could have consistent performance setting.

Naturally, `validate` would use this flag to force deep clones, to fix #76

**Comment by felixfbecker** [comment]
Wasn't there a `mutateDocument` flag in #167 that controlled cloning?

**Comment by alshakero** [comment]
@felixfbecker that controlled mutating the document, not the patch.

**Comment by baranga** [comment]
I rebased to `1.2.0`.

Benchmark results for `master`:
```
jsonpatch-core
==================
Benchmark results:
------------------
add operation
 Ops/sec: 2433304 ±3.91% Ran 143280 times in 0.059 seconds.
remove operation
 Ops/sec: 1253708 ±4.72% Ran 75032 times in 0.060 seconds.
replace operation
 Ops/sec: 2163807 ±0.96% Ran 115865 times in 0.054 seconds.
move operation
 Ops/sec: 225736 ±2.84% Ran 12869 times in 0.057 seconds.
copy operation
 Ops/sec: 219653 ±2.11% Ran 17094 times in 0.078 seconds.
test operation
 Ops/sec: 627258 ±1.26% Ran 33953 times in 0.054 seconds.
===================


jsonpatch-duplex
==================
Benchmark results:
------------------
add operation
 Ops/sec: 2216777 ±3.73% Ran 126016 times in 0.057 seconds.
remove operation
 Ops/sec: 1293096 ±1.56% Ran 70497 times in 0.055 seconds.
replace operation
 Ops/sec: 2249331 ±4.32% Ran 129225 times in 0.057 seconds.
move operation
 Ops/sec: 209552 ±1.85% Ran 14934 times in 0.071 seconds.
copy operation
 Ops/sec: 190221 ±4.24% Ran 21322 times in 0.112 seconds.
test operation
 Ops/sec: 634637 ±3.24% Ran 36097 times in 0.057 seconds.
===================


==================
Benchmark results:
------------------
generate operation
 Ops/sec: 458102 ±1.80% Ran 26931 times in 0.059 seconds.
generate operation and re-apply
 Ops/sec: 312761 ±1.63% Ran 18045 times in 0.058 seconds.
compare operation
 Ops/sec: 185162 ±1.00% Ran 9789 times in 0.053 seconds.
compare operation same but deep objects
 Ops/sec: 85377053 ±2.94% Ran 4607522 times in 0.054 seconds.
===================

```

Benchmark results for `issues/76_deepClone_1.0`:
```
jsonpatch-core
==================
Benchmark results:
------------------
add operation
 Ops/sec: 1954932 ±3.05% Ran 110794 times in 0.057 seconds.
remove operation
 Ops/sec: 1531813 ±1.13% Ran 82094 times in 0.054 seconds.
replace operation
 Ops/sec: 2275656 ±1.55% Ran 123980 times in 0.054 seconds.
move operation
 Ops/sec: 208920 ±2.01% Ran 17151 times in 0.082 seconds.
copy operation
 Ops/sec: 197645 ±2.98% Ran 17511 times in 0.089 seconds.
test operation
 Ops/sec: 688633 ±0.78% Ran 36317 times in 0.053 seconds.
===================


jsonpatch-duplex
==================
Benchmark results:
------------------
add operation
 Ops/sec: 2398309 ±2.83% Ran 135543 times in 0.057 seconds.
remove operation
 Ops/sec: 1383951 ±0.99% Ran 74409 times in 0.054 seconds.
replace operation
 Ops/sec: 2401083 ±2.46% Ran 130346 times in 0.054 seconds.
move operation
 Ops/sec: 203750 ±4.95% Ran 15942 times in 0.078 seconds.
copy operation
 Ops/sec: 195157 ±1.74% Ran 18125 times in 0.093 seconds.
test operation
 Ops/sec: 668618 ±1.40% Ran 35651 times in 0.053 seconds.
===================


==================
Benchmark results:
------------------
generate operation
 Ops/sec: 473497 ±1.38% Ran 27619 times in 0.058 seconds.
generate operation and re-apply
 Ops/sec: 319196 ±3.64% Ran 18791 times in 0.059 seconds.
compare operation
 Ops/sec: 205137 ±0.92% Ran 10933 times in 0.053 seconds.
compare operation same but deep objects
 Ops/sec: 86886870 ±2.92% Ran 4769806 times in 0.055 seconds.
===================
```

**Comment by alshakero** [comment]
Would appreciate creating an additional benchmarks with `copyByValue` set to true. Other than that, LGTM.

@tomalec WDYT?

File: src/json-patch-duplex.js [hunk 1]
```diff
@@ -70,6 +70,31 @@ var jsonpatch;
...
+        // from here https://jsperf.com/deep-copy-vs-json-stringify-json-parse/25 (recursiveDeepCopy)
+        var clone;
+        var i;
+        if (typeof value !== 'object') {
+            return value;
+        }
+        if (!value) {
+            return value;
+        }
+        if ('[object Array]' === Object.prototype.toString.apply(value)) {
```
Comment (felixfbecker) [inline src/json-patch-duplex.js]: this should use the `isArray` helper

File: src/json-patch-duplex.js [hunk 2]
```diff
@@ -70,6 +70,31 @@ var jsonpatch;
...
+            clone = [];
+            for (i = 0; i < value.length; i += 1) {
+                clone[i] = _deepClone(value[i]);
+            }
+            return clone;
+        }
+        clone = {};
+        for (i in value) {
+            if (value.hasOwnProperty(i)) {
```
Comment (felixfbecker) [inline src/json-patch-duplex.js]: `Object.keys()` is faster afaik

File: src/json-patch-duplex.js [hunk 3]
```diff
@@ -70,6 +70,31 @@ var jsonpatch;
...
+    function _deepClone(value) {
+        // from here https://jsperf.com/deep-copy-vs-json-stringify-json-parse/25 (recursiveDeepCopy)
+        var clone;
+        var i;
+        if (typeof value !== 'object') {
+            return value;
+        }
+        if (!value) {
+            return value;
+        }
```
Comment (felixfbecker) [inline src/json-patch-duplex.js]: these two ifs can be combined to `if (typeof value !== 'object' || value === null)`

File: src/json-patch-duplex.js [hunk 4]
```diff
@@ -587,7 +617,8 @@ var jsonpatch;
      * or just be undefined
      * @deprecated
      */
-    function apply(document, patch, validateOperation) {
+    function apply(document, patch, validateOperation, copyByValue) {
```
Comment (felixfbecker) [inline src/json-patch-duplex.js]: Imo deprecated API should be feature-locked - `mutateDocument` is not configurable either

File: src/json-patch-duplex.ts [hunk 5]
```diff
@@ -906,7 +938,7 @@ module jsonpatch {
       }
       if (document) {
         document = JSON.parse(JSON.stringify(document)); //clone document so that we can safely try applying operations
```
Comment (alshakero) [inline src/json-patch-duplex.ts]: Please use `_deepClone` helper here.

**Comment by alshakero** [comment]
@tomalec great idea! I just lean to doing that internally (cloning the whole patch) when the flag is true instead of documenting it. 

This will mean a very minimal change in the code (1-5 lines I guess) and would serve the purpose perfectly.

And can we settle this ASAP please? 🙏 I fixed https://github.com/Starcounter-Jack/JSON-Patch/issues/178 but didn't push to avoid @baranga rebasing yet one more time.

File: README.md [hunk 6]
```diff
@@ -178,6 +178,9 @@ Applies `patch` array on `obj`.
 
 An invalid patch results in throwing an error (see `jsonpatch.validate` for more information about the error object).
 
+It modifies the `document` object and `patch` - it gets the values by reference.
+If you would like to avoid touching your values, clone them: `jsonpatch.apply(document, jsonpatch.deepClone(patch))`.
```
Comment (alshakero) [inline README.md]: should be `applyPatch` in the example.

File: README.md [hunk 7]
```diff
@@ -199,6 +202,9 @@ Applies single operation object `operation` on `document`.
 - `validateOperation` Whether to validate the operation, or to pass a validator callback
 - `mutateDocument` Whether to mutate the original document or clone it before applying
 
+It modifies the `document` object and `operation` - it gets the values by reference.
+If you would like to avoid touching your values, clone them: `jsonpatch.apply(document, jsonpatch.deepClone(operation))`.
```
Comment (alshakero) [inline README.md]: I guess it should be `applyOperation`. Not being pedantic but `apply` is deprecated.

File: src/json-patch-duplex.d.ts [hunk 8]
```diff
@@ -115,6 +124,9 @@ declare module jsonpatch {
     /**
      * Apply a full JSON Patch array on a JSON document.
      * Returns the {newDocument, result} of the patch.
+     * It modifies the `document` object and `patch` - it gets the values by reference.
+     * If you would like to avoid touching your values, clone them:
+     * `jsonpatch.apply(document, jsonpatch.deepClone(patch))`.
```
Comment (alshakero) [inline src/json-patch-duplex.d.ts]: Same as above.

File: src/json-patch-duplex.ts [hunk 9]
```diff
@@ -710,6 +715,9 @@ module jsonpatch {
   /**
    * Apply a full JSON Patch array on a JSON document.
    * Returns the {newDocument, result} of the patch.
+   * It modifies the `document` object and `patch` - it gets the values by reference.
+   * If you would like to avoid touching your values, clone them:
+   * `jsonpatch.apply(document, jsonpatch.deepClone(patch))`.
```
Comment (alshakero) [inline src/json-patch-duplex.ts]: Same as above.

**Comment by alshakero** [comment]
Thank you so much for your persistence and patience @baranga! Merging..

**Pull request merged**